[rocprofiler-compute] Use native tool for counter collection

vedithal-amd opened the pull request:
## Motivation

Use native tool for counter collection

* Add native counter collection tool which uses rocprofiler-sdk C++ library public API to get counter collection data
    * This is enabled by default, unless --no-native-tool option is provided or ROCPROF=rocprofv3 env. var. is provided
    * This tool is only supported for ROCm version >=7.x.x
    * This tool is not supported for attach/detach scenario
* Build native tool shared object during build time
* If using rocprof-compute without building then runtime compilation of native tool shared object is performed
* rocprofiler-sdk tools is still used for services other than counter collection and data collected by native tool is merged into the rocpd/csv output of rocprofiler-sdk tool

* Make `rocpd` choice the default choice for `--format-rocprof-output` option
    * If `rocpd` public API from rocprofiler-sdk library is not present, then fallback to `csv` choice
    * In this case only `pmc_perf.csv` is written in workload folder instead of multiple `csv` files for each profiling run
* Remove `json` choice from `--format-rocprof-output` option since it functions identical to `csv` option

* Rename option `--rocprofiler-sdk-library-path` to `--rocprofiler-sdk-tool-path` since we LD_PRELOAD the
  rocprofiler-sdk tool shared object and not the rocprofiler-sdk library shared object

* Fix the meaning of `--dispatch` option in `profile` mode to mention dispatch iteration filtering instead of dispatch id filtering
    * --dispatch option in analyze mode does dispatch id filtering

* Move standalone binary creation logic from cmake file to docker file

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Technical Details

Explained in code comments

<!-- Explain the changes along with any relevant GitHub links. -->

## Test Plan

Ran all tests on MI 200 and MI 300 and ensure 100% pass rate
Fix formatting of code

<!-- Explain any relevant testing done to verify this PR. -->

## Test Result

All tests passing

<!-- Briefly summarize test outcomes. -->

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

vedithal-amd committed:
Use native tool for counter collection

* Add native counter collection tool which uses rocprofiler-sdk C++
  library public API to get counter collection data
    * This is enabled by default, unless --no-native-tool option is
      provided or ROCPROF=rocprofv3 env. var. is provided
    * This tool is only supported for ROCm version >=7.x.x
    * This tool is not supported for attach/detach scenario
* Build native tool shared object during build time
* If using rocprof-compute without building then runtime compilation of
  t push native tool shared object is performed
* rocprofiler-sdk tools is still used for services other than counter
  collection and data collected by native tool is merged into the
  rocpd/csv output of rocprofiler-sdk tool

* Make `rocpd` choice the default choice for `--format-rocprof-output`
  option
    * If `rocpd` public API from rocprofiler-sdk library is not present,
      then fallback to `csv` choice
    * In this case only `pmc_perf.csv` is written in workload folder
      instead of multiple `csv` files for each profiling run
* Remove `json` choice from `--format-rocprof-output` option since it
  functions identical to `csv` option

* Rename option `--rocprofiler-sdk-library-path` to
  `--rocprofiler-sdk-tool-path` since we LD_PRELOAD the
  rocprofiler-sdk tool shared object and not the rocprofiler-sdk library
shared object

* Fix the meaning of `--dispatch` option in `profile` mode to mention
  dispatch iteration filtering instead of dispatch id filtering
    * --dispatch option in analyze mode does dispatch id filtering

* Move standalone binary creation logic from cmake file to docker file

vedithal-amd committed:
fix native counter collection tool during attach/detach

vedithal-amd committed:
improve logging

vedithal-amd committed:
fix attach detach with native tool

vedithal-amd committed:
fix attach detach with native tool

vedithal-amd committed:
do not support attach/detach in native tool

vedithal-amd committed:
Update changelog

feizheng10 commented on projects/rocprofiler-compute/CMakeLists.txt:
```diff
@@ -599,42 +610,6 @@ add_custom_target(
         "src/${PACKAGE_NAME},cmake/Dockerfile,cmake/rocm_install.sh,docker/docker-entrypoint.sh,src/rocprof_compute_analyze/convertor/mongodb/convert"
 )
 
-# Standalone binary creation
```
I feel like we should keep standalonebinary both in cmake and docker build.
In another word, we should allow any one build standalonebinary without docker if they want

feizheng10 replied:
@cfallows-amd , what do you think?

vedithal-amd replied:
standalone binary creation using cmake is added

feizheng10 commented on projects/rocprofiler-compute/src/utils/utils.py:
```diff
@@ -1011,72 +1019,126 @@ def pc_sampling_prof(
 
-    if rocprof_output == "json":
-        results_files_json = glob.glob(f"{workload_dir}/out/pmc_1/*/*.json")
+    # Convert native counter collection csv to rocprofv3 counter collection csv format
```
Could you explain a bit more why do we need to do conversion here?
In case it might fail, should we try catch here?

vedithal-amd replied:
We already have logic to process rocprofiler-sdk tool counter collection csv to pmc_perf.csv
When using native tool for counter collection, the counter collection csv has to be converted to rocprofiler-sdk tool format to leverage this logic.

If it fails, then we should error out, we cannot silently catch it since that will result in no counter collection data!

feizheng10 replied:
I'm not saying "silently catch". I mean catch the error properly and error out

feizheng10 commented:
Using native tool as default will force whoever uses attach feature with --no-native-tool, right?

abchoudh-amd commented on projects/rocprofiler-compute/src/lib/rocprofiler_compute_tool.cpp:
```diff
@@ -0,0 +1,682 @@
+// input string. By using the iterators it finds the
+// window in the string which contains only the kernel name.
+// For example 'Foo<int, float>::foo(a[], int (int))' -> 'foo'
+std::string truncate_name(std::string_view name) {
```
The helper functions can be moved to a separate file (e.g: helper.cpp and helper.h). 
These would be useful if other features are added in future.

abchoudh-amd commented on projects/rocprofiler-compute/src/lib/rocprofiler_compute_tool.cpp:
```diff
@@ -0,0 +1,682 @@
+  return std::string{name.substr(rend - rit, rit - rbeg)};
+}
+
+std::string cxa_demangle(std::string_view _mangled_name, int *_status) {
```
Same as above

abchoudh-amd commented on projects/rocprofiler-compute/src/lib/rocprofiler_compute_tool.cpp:
```diff
@@ -0,0 +1,682 @@
+    return;
+
+  // get counters to collect
+  std::set<std::string> counters_to_collect;
```
The counter collection configuration generation logic can be moved to a separate function to make the code modular. 

Since, this is called only once per agent, it would be easier to understand the code logic; tldr: enhance readability.

abchoudh-amd commented on projects/rocprofiler-compute/src/lib/rocprofiler_compute_tool.cpp:
```diff
@@ -0,0 +1,682 @@
+// Tool data struct, now includes a vector of counter_info_record_t
+struct tool_data_t {
+  std::mutex mut{};
+  std::ostream *output_stream{nullptr};
```
Smart pointer can be used instead of raw pointer.

abchoudh-amd commented on projects/rocprofiler-compute/src/lib/rocprofiler_compute_tool.cpp:
```diff
@@ -0,0 +1,682 @@
+
+
+tool_data_t* create_tool_data(rocprofiler_client_id_t* id) {
+  auto *tool_data = new tool_data_t{};
```
Smart pointer would be safer.

vedithal-amd committed:
add standalone binary creation functionality in cmake

vedithal-amd commented:
> Using native tool as default will force whoever uses attach feature with --no-native-tool, right?

If user uses option --attach-pid it will automatically not use native tool, so --no-native-tool is implicit and does not have to be explicitly provided

vedithal-amd committed:
address review comments

vedithal-amd committed:
address review comments

vedithal-amd committed:
fix formatting

vedithal-amd committed:
address review comments

vedithal-amd commented:
/AzurePipelines run rocm-ci-caller

vedithal-amd commented:
/AzurePipelines run rocm-ci-caller

vedithal-amd commented:
/AzurePipelines run rocm-ci-caller

vedithal-amd commented:
/AzurePipelines run rocm-ci-caller

cfallows-amd committed:
Adding paths for cmake to search. Also updated min. cmake requirement to 3.21 as this was when hip was supported.

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

cfallows-amd committed:
Update hip compiler ID check, sometimes comes up as Clang, sometimes ROCMClang- depends on setup.
Updated formatting.

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

cfallows-amd committed:
RHEL8.10 unable to compile due to defaulting to old c++ version, need to force c++17

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -17,14 +24,19 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
 
 ### Changed
 
+* Default output format for underlying rocprofiler-sdk tool has been changed from csv to rocpd
```
```suggestion
* Default output format for the underlying ROCprofiler-SDK tool has been changed from ``csv`` to ``rocpd``.
```

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -17,14 +24,19 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
 ### Changed
 
+* Default output format for underlying rocprofiler-sdk tool has been changed from csv to rocpd
+  * If rocprofiler-sdk rocpd public library is not available, will fallback to csv format
```
```suggestion
  * If the ROCprofiler-SDK ``rocpd`` public library is not available, will fall back to ``csv`` format
```

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -17,14 +24,19 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
+* Default output format for underlying rocprofiler-sdk tool has been changed from csv to rocpd
+  * If rocprofiler-sdk rocpd public library is not available, will fallback to csv format
+
+* Option --rocprofiler-sdk-library-path has been changed to --rocprofiler-tool-library-path to better reflect the fact that we provide flexibility in choosing the path to rocprofiler-sdk tool and not the library.
```
```suggestion
* Option ``--rocprofiler-sdk-library-path`` has been changed to ``--rocprofiler-tool-library-path`` to better reflect the fact that we provide flexibility in choosing the path to ROCprofiler-SDK tool and not the library.
```

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -35,6 +47,12 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
 
 ### Resolved issues
 
+* Fixed the meaning of --dispatch option in profile mode in argparser to convey the fact that it control which iterations
```
```suggestion
* Fixed the description of ``--dispatch`` option in profile mode under ``argparser`` to clarify that it controls which kernel iterations are profiled, rather than which dispatch IDs are profiled.
```

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -35,6 +47,12 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
 ### Resolved issues
 
+* Fixed the meaning of --dispatch option in profile mode in argparser to convey the fact that it control which iterations
+  of the kernel to profile and not which dispatch ids to profile.
```
```suggestion
```

cfallows-amd committed:
Updating changelog per docs team recommendations

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -35,6 +47,12 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
 
+* Fixed the meaning of --dispatch option in profile mode in argparser to convey the fact that it control which iterations
+  of the kernel to profile and not which dispatch ids to profile.
+* The meaning of --dispatch option in analyze is still the same which is which dispatch ids to analyzer
```
```suggestion
* The description of the ``--dispatch`` option in analyze is still, which dispatch IDs to analyze.
```
Should be indented under the previous note.

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -35,6 +47,12 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
+* Fixed the meaning of --dispatch option in profile mode in argparser to convey the fact that it control which iterations
+  of the kernel to profile and not which dispatch ids to profile.
+* The meaning of --dispatch option in analyze is still the same which is which dispatch ids to analyzer
+* Fix the functioning of --dispatch option to act as 1-based index and ensure that correct kernel iterations are being
```
```suggestion
* Fixed the functioning of ``--dispatch`` option to act as a 1-based index and ensure that correct kernel iterations are being profiled.
```

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -35,6 +47,12 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
+  of the kernel to profile and not which dispatch ids to profile.
+* The meaning of --dispatch option in analyze is still the same which is which dispatch ids to analyzer
+* Fix the functioning of --dispatch option to act as 1-based index and ensure that correct kernel iterations are being
+  profiled
```
```suggestion
```

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -6,6 +6,13 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
+  * ROCm stack on the system is >= 7.0.0
+  * attaching to a process for profiling 
+
 * Add `--list-blocks <arch>` option to general options to list available IP blocks on specified arch (similar to `--list-metrics`), cannot be used with `--block`.
```
```suggestion
* `--list-blocks <arch>` option to general options. It lists the available IP blocks on the specified arch (similar to `--list-metrics`), however cannot be used with `--block`.
```

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -17,14 +24,19 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
+
+* Option --rocprofiler-sdk-library-path has been changed to --rocprofiler-tool-library-path to better reflect the fact that we provide flexibility in choosing the path to rocprofiler-sdk tool and not the library.
+
 * `-b/--block` accepts block alias(es) (See block aliases using command-line option `--list-blocks <arch>`).
```
```suggestion
* `-b/--block` accepts block alias(es). See block aliases using command-line option `--list-blocks <arch>`).
```
Can we provide more clarity on what we mean by (See block aliases using command-line option --list-blocks <arch>). Should we add a link to the document?

cfallows-amd replied:
```suggestion
* `-b/--block` accepts block alias(es). See block aliases using command-line option `--list-blocks <arch>`.
```
Updating suggestion to remove extra bracket

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -17,14 +24,19 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
+
 * `-b/--block` accepts block alias(es) (See block aliases using command-line option `--list-blocks <arch>`).
 
 * analysis configs yamls are now managed with the new config management workflow in `tools/config_management/`
```
```suggestion
* Analysis configs YAMLs are now managed with the new config management workflow in `tools/config_management/`.
```

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -6,6 +6,13 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
+
 * Add `--list-blocks <arch>` option to general options to list available IP blocks on specified arch (similar to `--list-metrics`), cannot be used with `--block`.
 
 * Added `config_delta/gfx950_diff.yaml` to analysis config yamls to track the revision between a gfx9 architecture against the latest supported architecture gfx950
```
```suggestion
* `config_delta/gfx950_diff.yaml` to analysis config YAMLs to track the revision between the gfx9xx GPUs against the latest supported gfx950 GPUs.
```

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -17,14 +27,19 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
-
 ### Removed
+
 * Removed `database` mode from `rocprofiler-compute`. This is to move our focus from grafana
```
```suggestion
* Removed `database` mode from ROCm Compute Profiler in favor of other visualization methods, rather than Grafana and MongoDB integration, such as the upcoming Analysis DB-based Visualizer.
```

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -17,14 +27,19 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
 ### Removed
+
 * Removed `database` mode from `rocprofiler-compute`. This is to move our focus from grafana
   and mongodb integration to other visualization methods such as:
```
```suggestion
```

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -17,14 +27,19 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
+
 * Removed `database` mode from `rocprofiler-compute`. This is to move our focus from grafana
   and mongodb integration to other visualization methods such as:
   * Analysis DB based Visualizer (upcoming)
```
```suggestion
```

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -4,8 +4,18 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
 
+* Native tool to perform counter collection using ROCprofiler-SDK public API
+  * Native tool is default for counter collection
+  * Native tool for counter collection will not be used in the following conditions:
```
```suggestion
  * Native tool for counter collection will not be used under the following conditions:
```

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -4,8 +4,18 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
 ### Added
 
+* Native tool to perform counter collection using ROCprofiler-SDK public API
+  * Native tool is default for counter collection
```
```suggestion
  * Native tool is now the default for counter collection.
```

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -4,8 +4,18 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
+* Native tool to perform counter collection using ROCprofiler-SDK public API
+  * Native tool is default for counter collection
+  * Native tool for counter collection will not be used in the following conditions:
+    * User provides a specific profiler through ``ROCPROF`` environment variable
```
```suggestion
    * Specific profiler is provided through the ``ROCPROF`` environment variable.
```

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -4,8 +4,18 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
+  * Native tool is default for counter collection
+  * Native tool for counter collection will not be used in the following conditions:
+    * User provides a specific profiler through ``ROCPROF`` environment variable
+    * ``--no-native-tool`` option is provided, forcing usage of default profiler
```
```suggestion
    * ``--no-native-tool`` option is provided, forcing usage of the default profiler.
```

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -4,8 +4,18 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
+  * Native tool for counter collection will not be used in the following conditions:
+    * User provides a specific profiler through ``ROCPROF`` environment variable
+    * ``--no-native-tool`` option is provided, forcing usage of default profiler
+    * System has ROCm version 7.0.0 or earlier. [Please confirm if it is for earlier (eg.6.4.1, 6.4.2) or later (eg. 7.0.1, 7.0.2)]
```
```suggestion
    * System has ROCm version 7.0.0 or earlier.
```
```suggestion
    * System has ROCm version 7.0.0 or earlier. 
```
Please confirm if it is for earlier (eg.6.4.1, 6.4.2) or later (eg. 7.0.1, 7.0.2) and update accordingly.

cfallows-amd replied:
```suggestion
```
Removing line and moving it up per recommendation in chat

prbasyal-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -4,8 +4,18 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
+    * User provides a specific profiler through ``ROCPROF`` environment variable
+    * ``--no-native-tool`` option is provided, forcing usage of default profiler
+    * System has ROCm version 7.0.0 or earlier. [Please confirm if it is for earlier (eg.6.4.1, 6.4.2) or later (eg. 7.0.1, 7.0.2)]
+    * Performing a live attach to a process for profiling
```
```suggestion
    * When performing a dynamic attach to a process for profiling.
```

prbasyal-amd reviewed (commented):
Editorial review feedback provided.

cfallows-amd commented on projects/rocprofiler-compute/CHANGELOG.md:
```diff
@@ -4,8 +4,18 @@ Full documentation for ROCm Compute Profiler is available at [https://rocm.docs.
+
 ### Added
 
+* Native tool to perform counter collection using ROCprofiler-SDK public API
```
```suggestion
* Native tool to perform counter collection using ROCprofiler-SDK public API. It is only supported for ROCm version 7.0.0 (and later).
```
Updating description per recommendation in chat

cfallows-amd committed:
Apply suggestions from code review to changelog

Co-authored-by: Pratik Basyal <pratik.basyal@amd.com>

cfallows-amd commented:
/AzurePipelines run rocm-ci-caller

cfallows-amd commented:
/AzurePipelines run rocm-ci-caller

vedithal-amd committed:
Do not required HIP complier to build native counter collection tool

vedithal-amd committed:
fix cmake

cfallows-amd committed:
gersemi formatting on latest cmake change

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

cfallows-amd committed:
ex ci updated dependencies to include rocprofiler-sdk, but cmake was still not capturing the path- there was a commit that added to the cmake_prefix_path entry that specified rocprof-sdk's cmake location ut was too specific for the search paths in find_package's config mode.
removing the cmake_prefix_path var and adding hints to find_package call instead, and specifying config mode so it knows how to construct the search paths

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

cfallows-amd committed:
gersemi run for formatting

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

cfallows-amd committed:
Still need prefix path, should not have been removed in last commit but does need to be shortened to just the rocm path to allow for find_package config mode to do the job

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

vedithal-amd committed:
include cstdint for uint32_t

cfallows-amd committed:
Run formatting on helper.cpp

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

vedithal-amd committed:
Remove rocm 7.2 release stuff from version and changelog and handle it in separate pr

cfallows-amd reviewed (approved):
Re-approving after changelog commits.

vedithal-amd committed:
fix version

vedithal-amd committed:
fix changelog

vedithal-amd committed:
fix changelog

cfallows-amd reviewed (approved):
lgtm with changelog merges.

cfallows-amd committed:
Merge branch 'develop' into users/vedithal/rocprofiler-compute-counter-collection

cfallows-amd commented:
/AzurePipelines run rocm-ci-caller

cfallows-amd committed:
run ruff formatter

Signed-off-by: Carrie Fallows <Carrie.Fallows@amd.com>

JeniferC99 commented:
/AzurePipelines run rocm-ci-caller

vedithal-amd committed:
fix rocprofiler-sdk attach so path

cfallows-amd reviewed (approved):
Reconfirmed path on recent rocm build, directory changed.. LGTM.
